[JSC] Add call_builtin wasm opcode

Constellation opened the pull request:
#### 841f77f24a8929071d7299612bfbc78177afa3e0
<pre>
[JSC] Add call_builtin wasm opcode
<a href="https://bugs.webkit.org/show_bug.cgi?id=249520">https://bugs.webkit.org/show_bug.cgi?id=249520</a>
rdar://103471989

Reviewed by Mark Lam.

This patch adds call_builtin Wasm opcode to save Wasm opcode space.
There are many Wasm opcodes which are just calling a function because of opcode&apos;s complexity.
We should represent it as a form of &quot;call_builtin&quot; instead and reduce # of opcodes.

In LLIntGenerator, we materialize call_builtin like what we materialize a normal call.
And instead of passing functionIndex, we pass builtinIndex. And then, call_builtin
slowpath offers a way to retrieve arguments from call&apos;s format so that we can pass arbitrary
number of values and we can return arbitrary number of values.

The definition of calling convension is following to wasm&apos;s call. And we use pre-defined
function signatures to materialize them correctly in LLIntGenerator.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h: Added.
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::addConstantWithoutPush):
(JSC::Wasm::LLIntGenerator::addConstant):
(JSC::Wasm::LLIntGenerator::addElemDrop):
(JSC::Wasm::LLIntGenerator::addTableSize):
(JSC::Wasm::LLIntGenerator::addTableCopy):
(JSC::Wasm::LLIntGenerator::addCallBuiltin):
(JSC::Wasm::LLIntGenerator::addCurrentMemory):
(JSC::Wasm::LLIntGenerator::addMemoryInit):
(JSC::Wasm::LLIntGenerator::addDataDrop):
(JSC::Wasm::LLIntGenerator::addMemoryFill):
(JSC::Wasm::LLIntGenerator::addMemoryCopy):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::jsrSize):
(JSC::LLInt::fprSize):
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::typeDefinitionForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:

Canonical link: <a href="https://commits.webkit.org/258047@main">https://commits.webkit.org/258047@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f5ab352c9e70f8b4bc8e3228c874093056ddf56

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100756 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9900 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110054 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170329 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10839 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/484 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93156 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107901 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106538 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8185 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91435 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90094 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22828 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91242 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3594 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24352 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87333 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1130 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3618 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29206 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9728 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43851 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90217 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5395 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20187 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2889 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -623,8 +626,8 @@ auto LLIntGenerator::callInformationForCaller(const FunctionSignature& signature
-    // FPRn
-    // ...
+    // FPRn                        |
+    // ...                         v stack-growth
```
nit: lets say "stack-growth towards lower memory" instead.  I was also confused for a second when I first read this code.  I think adding the explicit detail about which side is lower memory helps.

Constellation replied:
Added.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -715,6 +649,103 @@ WASM_SLOW_PATH_DECL(call_ref_no_tls)
+    auto takeGPR = [&]() -> Register& {
+        if (gprIndex != jsrSize()) {
+            unsigned cursor = gprIndex++;
+            return *(gprStart + cursor);
```
Why not just express this as `gprStart[gprIndex++]`?

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -715,6 +649,103 @@ WASM_SLOW_PATH_DECL(call_ref_no_tls)
+            return *(gprStart + cursor);
+        }
+        unsigned cursor = stackIndex++;
+        return *(stackStart + cursor);
```
Ditto: express as `stackStart[stackIndex++]`?

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -536,30 +536,26 @@ bool RTT::isSubRTT(const RTT& parent) const
-        thunkTypes[linearizeType(TypeKind::Void)] = sig.get();
-        m_typeSet.add(TypeHash { sig.releaseNonNull() });
-    }
+const TypeDefinition& TypeInformation::typeDefinitionForLLIntBuiltin(LLIntBuiltin builtin)
```
nit: let's call this `signatureForLLIntBuiltin` instead.

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -581,6 +582,8 @@ class TypeInformation {
 public:
     static TypeInformation& singleton();
 
+    static const TypeDefinition& typeDefinitionForLLIntBuiltin(LLIntBuiltin);
```
Rename to `signatureForLLIntBuiltin`.

Constellation replied:
Changed.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Add call_builtin wasm opcode
https://bugs.webkit.org/show_bug.cgi?id=249520
rdar://103471989

Reviewed by Mark Lam.

This patch adds call_builtin Wasm opcode to save Wasm opcode space.
There are many Wasm opcodes which are just calling a function because of opcode's complexity.
We should represent it as a form of "call_builtin" instead and reduce # of opcodes.

In LLIntGenerator, we materialize call_builtin like what we materialize a normal call.
And instead of passing functionIndex, we pass builtinIndex. And then, call_builtin
slowpath offers a way to retrieve arguments from call's format so that we can pass arbitrary
number of values and we can return arbitrary number of values.

The definition of calling convension is following to wasm's call. And we use pre-defined
function signatures to materialize them correctly in LLIntGenerator.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h: Added.
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::addConstantWithoutPush):
(JSC::Wasm::LLIntGenerator::addConstant):
(JSC::Wasm::LLIntGenerator::addElemDrop):
(JSC::Wasm::LLIntGenerator::addTableSize):
(JSC::Wasm::LLIntGenerator::addTableCopy):
(JSC::Wasm::LLIntGenerator::addCallBuiltin):
(JSC::Wasm::LLIntGenerator::addCurrentMemory):
(JSC::Wasm::LLIntGenerator::addMemoryInit):
(JSC::Wasm::LLIntGenerator::addDataDrop):
(JSC::Wasm::LLIntGenerator::addMemoryFill):
(JSC::Wasm::LLIntGenerator::addMemoryCopy):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::jsrSize):
(JSC::LLInt::fprSize):
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::typeDefinitionForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:

Canonical link: https://commits.webkit.org/258047@main
